controllers/imagerepository_notifications.go

```diff
@@ -0,0 +1,186 @@
+package controllers
+
+import (
+	"context"
+
+	imagerepositoryv1alpha1 "github.com/konflux-ci/image-controller/api/v1alpha1"
+	"github.com/konflux-ci/image-controller/pkg/quay"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func (r *ImageRepositoryReconciler) AddNotification(notification imagerepositoryv1alpha1.Notifications, imageRepository *imagerepositoryv1alpha1.ImageRepository) (imagerepositoryv1alpha1.NotificationStatus, error) {
+	quayNotification, err := r.QuayClient.CreateNotification(
+		r.QuayOrganization,
+		imageRepository.Spec.Image.Name,
+		quay.Notification{
+			Title:  notification.Title,
+			Event:  string(notification.Event),
+			Method: string(notification.Method),
+			Config: quay.NotificationConfig{
+				Url: notification.Config.Url,
+			},
+			EventConfig: quay.NotificationEventConfig{},
+		})
+
+	return imagerepositoryv1alpha1.NotificationStatus{
+		UUID:  quayNotification.UUID,
+		Title: notification.Title,
+	}, err
+}
+
+func (r *ImageRepositoryReconciler) HandleNotifications(ctx context.Context, imageRepository *imagerepositoryv1alpha1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx).WithName("HandleNotifications")
+
+	if imageRepository.Status.Notifications == nil {
+		// No status notifications to check
+		return nil
+	}
+	allNotifications, err := r.QuayClient.GetNotifications(r.QuayOrganization, imageRepository.Spec.Image.Name)
+	if err != nil {
+		if err = r.UpdateImageRepositoryStatus(ctx, imageRepository, "Couldn't retrieve all Quay notifications"); err != nil {
+			return err
+		}
+		return nil
+	}
+	for index := 0; index < len(imageRepository.Status.Notifications); index++ {
+		statusNotification := &imageRepository.Status.Notifications[index] // This way we can update the UUID if notification gets updated
+		existsInSpec := false
+		for _, notification := range imageRepository.Spec.Notifications {
```

Comment:
You can see here how the structure of Spec.Notifications and Status.Notifications is: https://github.com/konflux-ci/image-controller/pull/112#issue-2302454077

Only Status.Notifications have UUID, the piece required to interact with Quay API to do operations such as **Update** and **Delete**. At the end, we have to compare all Status.Notifications items with the ones in Spec.Notifications and when we do not have a match, we check if that's because a new item is present in Spec.Notifications and in that case we do the **Add** operation.

Since the UUID is present in Status.Notifications it makes more sense to me (and is more legible/structured) to compare that items with the ones in Spec.Notifications (+ we could have less loop iterations by starting with Status.Notifications items)

Reply:
Agree, thank you for the explanation.